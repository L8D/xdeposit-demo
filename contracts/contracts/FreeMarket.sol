// SPDX-License-Identifier: MIT
pragma solidity >=0.4.22 <0.9.0;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
import './AddressSet.sol';
import './Ownable.sol';

contract FreeMarket is Ownable {

  event Deposit(address indexed sender, uint256 amount, address token);

  address public busStop;
  
  AddressSet.Set supportedERC20Tokens;

  constructor() {
    owner = msg.sender;
  }

  function setBusStop(address newBusStop) public onlyOwner {
    busStop = newBusStop;
  }

  function getSupportedERC20Tokens() public view returns(address[] memory) {
    return supportedERC20Tokens.values;
  }

  function addSupportedERC20Token(address erc20TokenAddress) public onlyOwner {
    AddressSet.add(supportedERC20Tokens, erc20TokenAddress);
  }

  function removeSupportedERC20Token(address erc20TokenAddress) public onlyOwner {
    AddressSet.remove(supportedERC20Tokens, erc20TokenAddress);
  }

  function deposit(address erc20TokenAddress, uint256 amount) public {
    require(AddressSet.exists(supportedERC20Tokens, erc20TokenAddress), "token not supported");
    IERC20 token = IERC20(erc20TokenAddress);
    SafeERC20.safeTransferFrom(token, msg.sender, busStop, amount);
    emit Deposit(msg.sender, amount, erc20TokenAddress);
  }
}
