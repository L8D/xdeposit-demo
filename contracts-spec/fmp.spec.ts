import * as ethers from 'ethers'
import FreeMarket from '@fmp/contracts/build/contracts/FreeMarket.json'
import TricryptoPool from '@fmp/contracts/TricryptoPool.json'
import WrappedEther from '@fmp/contracts/WrappedEther.json'
import Parser from 'gherkin/dist/src/Parser'
import AstBuilder from 'gherkin/dist/src/AstBuilder'
import { IdGenerator } from '@cucumber/messages'
import { readFileSync } from 'fs'

const provider = new ethers.providers.JsonRpcProvider('http://127.0.0.1:8545')

const mnemonic = process.env.FMP_GANACHE_MNEMONIC

if (mnemonic == null) {
  throw new Error('missing FMP_GANACHE_MNEMONIC')
}

const hdNode = ethers.utils.HDNode.fromMnemonic(mnemonic)

const account = new ethers.Wallet(
  hdNode.derivePath("m/44'/60'/0'/0/1").privateKey,
  provider
)

const contractFactory = new ethers.ContractFactory(
  FreeMarket.abi,
  FreeMarket.bytecode,
  account
)

const tricrypto = new ethers.Contract(
  '0xD51a44d3FaE010294C616388b506AcdA1bfAAE46',
  TricryptoPool as any,
  account
)

const WETH_ADDRESS = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'
const weth = new ethers.Contract(WETH_ADDRESS, WrappedEther, account)

const USDT_ADDRESS = '0xdAC17F958D2ee523a2206206994597C13D831ec7'
const usdt = new ethers.Contract(USDT_ADDRESS, WrappedEther.filter(({name}) => name === 'balanceOf'), account)

const astBuilder = new AstBuilder(IdGenerator.incrementing())

const parser = new Parser(astBuilder)

const { feature } = parser.parse(
  readFileSync(`${__dirname}/fmp.feature`, 'utf8').toString()
)

describe('FreeMarket contract', () => {
  for (const { scenario } of feature.children) {
    it(scenario.name, async () => {
      const contract = await contractFactory.deploy()
      await contract.deployed()

      for (const step of scenario.steps) {
        if (step.text === 'I exchange 1 WETH for USDT') {
          const wethBalance = await weth.balanceOf(account.address)
            const [expected] = await tricrypto.functions.get_dy(2, 0, wethBalance, { gasLimit: 3000000000 })
            console.log(expected.toString())
            console.log(tricrypto)
          const approvalTx = await weth['approve'](
            tricrypto.address,
            wethBalance,
          )
          console.log('approvalTx', approvalTx)
          const approvalReceipt = await approvalTx.wait()
          console.log('approvalReceipt', approvalReceipt)
          console.log('approved')
          const tx = await tricrypto['exchange(uint256,uint256,uint256,uint256,bool)'](
            2, // weth
            0, // usdt
            wethBalance,
            1,
            false
          )
          await tx.wait()
        } else if (step.text === 'I exchange 1 ETH for WETH') {
          const tx = await weth.functions.deposit({ value: '100000000000000000' })
          await tx.wait()
          const wethBalance = (await weth.balanceOf(account.address)).toString()
          expect(wethBalance).not.toBe('0')
          // expect(wethBalance).toBe('10000000000000000')
        } else if (step.text === 'I should have some USDT') {
          const balance = await usdt.balanceOf(account.address)
          expect(balance.toString()).not.toBe('0')
        }
      }
    })
  }
})
