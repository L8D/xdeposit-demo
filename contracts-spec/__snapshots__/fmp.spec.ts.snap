// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeMarket contract I can deposit 1`] = `
Array [
  Object {
    "args": Array [
      "0x2Ba7ab10fC7f42E576C4E6092C772563eFFC8159",
      "0x389ABaC16AE9AB8A886E205Da04bDA01e4A1F20c",
      Object {
        "hex": "0x02faf080",
        "type": "BigNumber",
      },
    ],
    "name": "",
  },
  Object {
    "args": Array [
      "0x2Ba7ab10fC7f42E576C4E6092C772563eFFC8159",
      Object {
        "hex": "0x02faf080",
        "type": "BigNumber",
      },
      "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    ],
    "name": "Deposit",
  },
]
`;
